### PR TITLE
RDRP-913: Add check for frozen data before staging; validate the frozen csv schema.

### PR DESCRIPTION
**Pull request by cbrowning04**
## Pull Request submission

*Insert detailed bullet points about your changes here!*

*Insert any instructions to help the reviewer, e.g. "install new requirements from `requirements.txt`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/freezing/freezing_main.py [hunk 1]
```diff
@@ -91,3 +97,24 @@ def run_freezing(
 def read_frozen_csv():
     # new functionality
     pass
+
+# function ready for use
+def _add_last_frozen_column(
```
Comment (JenCheshire) [inline src/freezing/freezing_main.py:108]: Can this be run at line 59 on the updated snapshot for run_first_snapshot_of_results (we'll use it for updates in the apply freezing adds and amends process)

File: src/freezing/freezing_main.py [hunk 2]
```diff
@@ -43,6 +46,9 @@ def run_freezing(
     if not run_first_snapshot_of_results:
         FreezingLogger.info("Loading frozen data...")
         frozen_data_staged_csv = read_csv(frozen_data_staged_path)
```
Comment (JenCheshire) [inline src/freezing/freezing_main.py]: Pipeline isn't reading the frozen snapshot

**Review by JenCheshire — COMMENTED** [review]
Looking good so far - just two things to look at, so we the pipeline runs fully for first run and from frozen csv.  Thanks

File: src/freezing/freezing_main.py [hunk 3]
```diff
@@ -43,13 +46,17 @@ def run_freezing(
...
+            frozen_data_staged_path, "./config/frozen_data_staged_schema.toml"
+        )
         FreezingLogger.info(
             "Frozen data successfully read from {frozen_data_staged_path}"
         )
         return frozen_data_staged_csv
 
     if run_first_snapshot_of_results:
         updated_snapshot = main_snapshot.copy()
+        updated_snapshot = _add_last_frozen_column(updated_snapshot)
```
Comment (JenCheshire) [inline src/freezing/freezing_main.py]: Needs run ID

File: src/user_config.yaml [hunk 4]
```diff
@@ -56,7 +56,7 @@ hdfs_paths:
   ni_full_responses_path: "03_northern_ireland/2021/TEST_ni.csv"
   feather_path: "staging/feather"
   # Freezing data paths
-  frozen_data_staged_path: "02_freezing/frozen_data_staged"
+  frozen_data_staged_path: "02_freezing/frozen_data_staged/2023_FROZEN_staged_BERD_full_responses_24-08-05_v817.csv"
```
Comment (JenCheshire) [inline src/user_config.yaml]: It's not reading it in - was it working for you?

File: src/freezing/freezing_main.py [hunk 5]
```diff
@@ -84,24 +88,24 @@ def run_freezing(
...
         FreezingLogger.info("Outputting frozen data file.")
         tdate = datetime.now().strftime("%y-%m-%d")
         survey_year = config["years"]["survey_year"]
         filename = (
             f"{survey_year}_FROZEN_staged_BERD_full_responses_{tdate}_v{run_id}.csv"
         )
-        write_csv(f"{frozen_data_staged_path}/{filename}", updated_snapshot)
-
-    return updated_snapshot
+        write_csv(f"{frozen_data_staged_output_path}/{filename}", prepared_frozen_data)
```
Comment (cbrowning04) [inline src/freezing/freezing_main.py]: os.path.join safer here?

**Review by cbrowning04 — COMMENTED** [review]
Just one small thing @JenCheshire , but all looks fine. Good stuff

**Comment by cbrowning04** [comment]
Happy for this to be merged :) @JenCheshire

**Review by JenCheshire — APPROVED** [review]

**Pull request merged**